Add setters to the StashNotifier.DescriptorImpl class

alexchiri opened the pull request:
In order to be able to configure this plugin using the [Configuration As Code plugin](https://github.com/jenkinsci/configuration-as-code-plugin), the StashNotifier.DescriptorImpl needs to have setters or public fields. Using YAML to configure Jenkins is pretty nice and this would enable this to work with this plugin as well.

alexchiri committed:
Add setters to the StashNotifier.DescriptorImpl class

scaytrase commented:
All or most of this properties could be set while construction. While should we allow dynamically change the settings during lifecycle. Isn't it enough to pass YAML configuration to constructor in some way?

alexchiri commented:
@scaytrase Unfortunately, at least as far as [I can see in the plugin implementation](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/casc/BaseConfigurator.java#L50), it works by discovering fields and setters to determine what properties can be set from the YAML. They have some guidelines for plugin developers [here](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/PLUGINS.md).  There can be some custom implementation be done on the configuration-as-code plugin side to handle different ways of configuring plugins, but I felt since stash notifier has quite a basic config, that would be a bit of an overkill.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -611,34 +611,66 @@ public String getStashRootUrl() {
         }
 
+        public void setStashRootUrl(String stashRootUrl) {
+			this.stashRootUrl = stashRootUrl;
```
Please use correct indentation for all new code.
Here and in other places spaces & tabs are currently mixed up.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -611,34 +611,66 @@ public String getStashRootUrl() {
             }
         }
 
+        public void setStashRootUrl(String stashRootUrl) {
```
Please annotate all setters with `@DataboundSetter`.
See **Step 1** in https://github.com/jenkinsci/configuration-as-code-plugin/blob/configuration-as-code-1.8/docs/PLUGINS.md.

darxriggs reviewed (commented):
Please also use data-binding via `request.bindJson(this, json)` inside `configure(...)`.
See **Step 5** in https://github.com/jenkinsci/configuration-as-code-plugin/blob/configuration-as-code-1.8/docs/PLUGINS.md.

There should also be two tests - one for configuring Jenkins via YAML and one for exporting a configured Jenkins to YAML.
See **How to test?** in https://github.com/jenkinsci/configuration-as-code-plugin/blob/configuration-as-code-1.8/docs/PLUGINS.md.

Note: I am not a maintainer of this plugin. But I already have migrated some other plugins to be compatible with JCasC.

alexchiri commented:
@darxriggs Thank you for the feedback and review. Will work further on the PR soon.

scaytrase commented:
@alexchiri as soon I have no views on architectural future of this plugin version (only some thoughts on 2.x version), I think we can accept this PR after suggested impovements

alexchiri committed:
Merge branch 'release/1.x' into add-setters-to-stash-notifier-descriptor

alexchiri committed:
Update version of jenkins-test-harness to the latest version (useful to run the tests in IntelliJIDEA)

alexchiri committed:
Update parent plugin version, Jenkins version and Java level version, to match JCasC plugin.
Add JCasC plugin test dependency to be able to test config as code functionality.
Add matrix-project plugin as a test dependency, since it is required in tests by the JenkinsRule.
Fix spacing and add @DataBoundSetter annotations to setters.
Rewrite configure method and test to use req.bindJSON.
Add tests to verify that configure using yaml and export to yaml work.

darxriggs commented on pom.xml:
```diff
@@ -3,20 +3,20 @@
         <changelist>-SNAPSHOT</changelist>
         <!-- Baseline Jenkins version you use to build the plugin. Users must have this version or newer to run. -->
-        <jenkins.version>1.625.3</jenkins.version>
+        <jenkins.version>2.60.3</jenkins.version>
```
I suggest  to reduce it to the following snippet:
```
<properties>
    <revision>1.18</revision>
    <changelist>-SNAPSHOT</changelist>
    <jenkins.version>2.60.3</jenkins.version>
    <java.level>8</java.level>
</properties>
```
The rest is already covered by the parent pom nowadays.
(This can be verified with `mvn help:effective-pom`.)

darxriggs commented on pom.xml:
```diff
@@ -3,20 +3,20 @@
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>3.10</version>
+        <version>3.36</version>
```
If you like, there is already `3.40` - see https://github.com/jenkinsci/plugin-pom/releases.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -637,34 +626,73 @@ public String getStashRootUrl() {
             return projectKey;
         }
 
+        public void setProjectKey(String projectKey) {
```
It seems you missed a `@DataBoundSetter` here.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -637,34 +626,73 @@ public String getStashRootUrl() {
 
+        @DataBoundSetter
+        public void setStashRootUrl(String stashRootUrl) {
+            this.stashRootUrl = stashRootUrl;
```
For Strings like this one I suggest to use something like Apache Commons Lang [StringUtils::trimToNull](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#trimToNull-java.lang.String-) or Jenkins [Util::fixEmptyAndTrim](https://javadoc.jenkins.io/hudson/Util.html#fixEmptyAndTrim-java.lang.String-). This improves user experience by better handling strings with leading/trailing spaces.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -637,34 +626,73 @@ public String getStashRootUrl() {
         }
 
+        public void setProjectKey(String projectKey) {
+            this.projectKey = projectKey;
```
See comment about `setStashRootUrl()` on trimming.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -637,34 +626,73 @@ public String getStashRootUrl() {
 
+        @DataBoundSetter
+        public void setCredentialsId(String credentialsId) {
+            this.credentialsId = credentialsId;
```
See comment about `setStashRootUrl()` on trimming.

darxriggs commented on src/test/java/org/jenkinsci/plugins/stashNotifier/ConfigAsCodeTest.java:
```diff
@@ -0,0 +1,60 @@
+        assertThat(stashNotifierConfig.isPrependParentProjectKey(), equalTo(true));
+        assertThat(stashNotifierConfig.isDisableInprogressNotification(), equalTo(true));
+        assertThat(stashNotifierConfig.isConsiderUnstableAsSuccess(), equalTo(true));
+    }
```
Verifying `stashRootUrl` is missing here.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -720,26 +748,20 @@ public boolean configure(
-            stashRootUrl = formData.getString("stashRootUrl");
-            ignoreUnverifiedSsl = formData.getBoolean("ignoreUnverifiedSsl");
-            includeBuildNumberInKey = formData.getBoolean("includeBuildNumberInKey");
+            this.stashRootUrl = null;
```
I suggest to keep the order consistent for easier reading/comparison:
* field declarations in `StashNotifier` (maybe)
* setters/getters in `StashNotifier` (maybe)
* default values (lines 751-758) in `StashNotifier::configure`
* assertThat calls (lines 27-33) in `ConfigAsCodeTest::should_support_jcasc_from_yaml`
* setter calls (lines 40-47) in `ConfigAsCodeTest::should_support_jcasc_to_yaml`

darxriggs commented on src/test/resources/org/jenkinsci/plugins/stashNotifier/ConfigAsCodeTest/configuration-as-code.yml:
```diff
@@ -0,0 +1,10 @@
+    includeBuildNumberInKey: true
+    prependParentProjectKey: true
+    projectKey: "JEN"
+    stashRootUrl: "https://bitbucket.com"
```
Even if it's only an example value, I suggest to use something like `https://my.company.intranet/bitbucket`. This more clearly states that this plugin is about "Bitbucket Server" and not "Bitbucket Cloud". Also see #225 for a cleanup regarding this.

darxriggs commented on src/test/java/org/jenkinsci/plugins/stashNotifier/DescriptorImplTest.java:
```diff
@@ -73,21 +73,13 @@ public void setUp() throws Exception {
-        assertThat(desc.isIncludeBuildNumberInKey(), is(true));
-        assertThat(desc.isPrependParentProjectKey(), is(true));
-        assertThat(desc.isApplicable(AbstractProject.class), is(true));
+        verify(request).bindJSON(desc, json);
```
The question is, if just verifying a `bindJSON` call is sufficient for you.
To be really sure, values should still be verified (using a `StaplerRequest` implementation/mock that implements `bindJSON` if possible). Otherwise if you for example missed a setter for a field, this would not be caught here and JCasC and UI based configuration would be broken.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -311,10 +308,6 @@ private boolean perform(Run<?, ?> run,
     private String getRootUrl() {
         Jenkins instance = Jenkins.getInstance();
 
-        if (null == instance) {
```
I think this change is ok.
`Jenkins::getInstance` is now annotated with `Nonnull` and this code only runs when Jenkins is completely initialized.

alexchiri replied:
Yup, and also spotbugs doesn't give you much choice on this. Without this change and the one below, `mvn install` wouldn't complete successfully.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -531,10 +524,6 @@ public boolean isTrusted(X509Certificate[] chain, String authType)
 
     private void configureProxy(HttpClientBuilder builder, URL url) {
         Jenkins jenkins = Jenkins.getInstance();
-        if (jenkins == null) {
```
I think this change is ok.
See comment on `getRootUrl()`.

darxriggs reviewed (changes requested):
Thanks for the quick update.
I had another look at the new changes. It already looks pretty good.
See my comments for some required changes.

alexchiri committed:
Make updates based on feedback from @darxriggs

Update parent version to latest.
Remove Maven properties that are inherited anyway from parent pom.
Order plugin attributes in the same way everywhere.
Trim String attributes.
Add forgotten @DataBoundSetter.
Update example value for stashRootUrl.
Implement more rigorous unit test for configure method.

alexchiri commented:
Thanks again for the detailed feedback and advice on this @darxriggs. Learned some things and feel I'm getting better at it. 😊 

Please let me know if anything else sticks out!

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -629,42 +615,82 @@ public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Item project) {
+            this.prependParentProjectKey = prependParentProjectKey;
+        }
+
+        public String getStashRootUrl() {
```
Now that `trimToNull()` is used in `setStashRootUrl()` this can simply return the `stashRootUrl` value (like all other getters).

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -629,42 +615,82 @@ public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Item project) {
+        }
+
         public String getProjectKey() {
             return projectKey;
```
You moved the field `projectKey` before the field `stashRootUrl` but the setter/getter order doesn't match this. But it doesn't really matter.

alexchiri committed:
Remove null check on getter.

alexchiri committed:
Update order of getters/setters.

darxriggs commented on src/test/java/org/jenkinsci/plugins/stashNotifier/DescriptorImplTest.java:
```diff
@@ -74,20 +77,30 @@ public void testConfigure() throws Descriptor.FormException {
         //given
         doNothing().when(desc).save();
 
+        ServletContext servletContext = PowerMockito.mock(ServletContext.class);
```
If possible try not to use PowerMockito but only Mockito. You can have a look in the internet why to avoid PowerMock.

PowerMockito is mainly required to mock constructors and static methods. That's often the case for legacy code that was written without testability in mind. At first sight that's not the case here.

darxriggs commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -584,14 +570,14 @@ public DescriptorImpl getDescriptor() {
+        private boolean disableInprogressNotification;
         private boolean ignoreUnverifiedSsl;
         private boolean includeBuildNumberInKey;
-        private String projectKey;
```
@scaytrase I am wondring why `projectKey` is a field in `StashNotifier.DescriptorImpl`. It is not visible in the UI and was never part of `global.jelly`. Shouldn't it be removed from `StashNotifier.DescriptorImpl` and only be a field in `StashNotifier`?

scaytrase replied:
Not sure for it really, may be a copy-paste coincident. We can try to remove it, cannot image when project key should be declared on global level

scaytrase replied:
if it's non blocking for this PR we can try it as a separate change

alexchiri committed:
Replace PowerMockito with Mockito in unit test.

darxriggs commented:
Apart from the `projectKey` field issue, which is to be decided by @scaytrase, this pull request now looks good to me.
I suggest to squash all the commits into one before merging.

scaytrase commented:
thank you @alexchiri ! (and @darxriggs for comments, for sure)

proski commented on pom.xml:
```diff
@@ -143,6 +135,18 @@
         </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>matrix-project</artifactId>
```
Why was the matrix project dependency added? The commit description makes no mention of it. I don't see the matrix plugin being used in the code. I see that it's only for testing, but I still don't see those new tests.

darxriggs replied:
That's a `test` scope dependency required by the new tests.

alexchiri replied:
Sorry, I missed to add that. I added the Matrix project test dependency because the JenkinsRule requires it and it has been removed from core since 1.561. I added the JenkinsRule in the test that verifies the configuration as code setup, according to JCasC documentation. (import and export yaml). (this is the issue where I found more information about the NoClassDefFoundError I got while using the JenkinsRule: https://issues.jenkins-ci.org/browse/JENKINS-27826)

proski replied:
Thank you for the explanation!

proski commented on src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:
```diff
@@ -21,10 +21,7 @@
-import hudson.FilePath;
-import hudson.Launcher;
-import hudson.ProxyConfiguration;
+import hudson.*;
```
I don't think wildcards make the code better. There are plenty of tools to keep an up-to-date list of used imports.

darxriggs replied:
For me, using wildcard imports or not is a project decision. There are as always advantages and disadvantages. And nowadays tools should be able to handle both.

proski replied:
I agree on all counts, I just wanted to make sure it was an intentional change and not somebody's sloppiness that everybody missed.

scaytrase replied:
I'm not in general a Java developer and I adopted (ocassionaly, huh) this plugin with worse code conventions than it have today, but there is still no CSC strictly defined here today. We can adopt one, I think, if it make life easier.

proski reviewed (commented):
It was surprising to find this large commit in the commit history with a description that makes it sound like a minor change.

darxriggs commented:
@scaytrase I am looking forward for this to be released. :-)

darxriggs commented:
Just as reference, there is also [JENKINS-56899](https://issues.jenkins-ci.org/browse/JENKINS-56899) tracking this.
